### PR TITLE
feat(linter): eslint-plugin-unicorn/number-literal-case

### PR DESCRIPTION
**Pull request by haocheng6**
This PR implements the [eslint-plugin-unicorn/number-literal-case](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/number-literal-case.md) rule.

Tests taken from https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/test/number-literal-case.mjs. (Test cases that rely on a Vue parser are removed.)

Related issue: #684

### CODE REVIEW

File: crates/oxc_linter/src/rules/unicorn/number_literal_case.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+impl Message {
+    fn details(&self) -> (&'static str, &'static str) {
+        match self {
+            Self::UppercasePrefix => ("Unexpected number literal prefix in uppercase.", "Use lowercase for number literal prefixes (`0b`, `0o`, and `0x`)."),
+            Self::UppercaseExponentialNotation => ("Unexpected exponential notation in uppercase.", "Use lowercase for `e` in exponential notations."),
+            Self::LowercaseHexadecimalDigits => ("Unexpected hexadecimal digits in lowercase.", "Use uppercase for hexadecimal digits."),
+            Self::UppercasePrefixAndLowercaseHexadecimalDigits => ("Unexpected number literal prefix in uppercase and hexadecimal digits in lowercase.", "Use lowercase for number literal prefixes (`0b`, `0o`, and `0x`) and uppercase for hexadecimal digits.")
+        }
+    }
+}
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/number_literal_case.rs:146]: shall we make the diagnostic an enum instead?


https://github.com/oxc-project/oxc/blob/db5417f9a954b18a7dfd867b1ee9d7d7b7655db2/crates/oxc_linter/src/rules/eslint/valid_typeof.rs#L13-L21

File: crates/oxc_linter/src/rules/unicorn/number_literal_case.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    correctness
+);
+
+impl Rule for NumberLiteralCase {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let (raw_literal, raw_span) = match node.kind() {
+            AstKind::NumberLiteral(number) => (number.raw, number.span),
+            AstKind::BigintLiteral(number) => {
+                let span = number.span;
+                (&ctx.source_text()[span.start as usize..span.end as usize], span)
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/number_literal_case.rs:59]: does this work?
```
(span.source_text(ctx.source_text(), span) 
```

**Review by camc314 — COMMENTED** [review]
looking good!

File: crates/oxc_linter/src/rules/unicorn/number_literal_case.rs (hunk 1 above)
Comment (haocheng6) [reply crates/oxc_linter/src/rules/unicorn/number_literal_case.rs]: Thanks for the suggestion! Updated in 2a58b6b77859ad28891c7036d007b834f93dee5f. I slightly improved the help message for some cases.

BTW, I also changed the category of this rule from "correctness" to "style" in 2632f65ce93f6f8fe8b2acdc1c61d0ae615a87dc.

File: crates/oxc_linter/src/rules/unicorn/number_literal_case.rs (hunk 2 above)
Comment (haocheng6) [reply crates/oxc_linter/src/rules/unicorn/number_literal_case.rs]: Good catch - updated in 2a58b6b77859ad28891c7036d007b834f93dee5f.

**Review by camc314 — APPROVED** [review]

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Thank you!

**Pull request merged**